### PR TITLE
ECDSA signing and verification

### PR DESCRIPTION
**Pull request by ewynx**
# ECDSA
This PR contains signing and verification.

## Testing
The tests have been moved to the already existing `tests` folder. For each separate bulk of functionality there is a separate testfolder, for example all curve arithmetic tests can be found in `testing/p256_tests/curve_test`. Test instructions are in  `testing/p256_tests/README.md`. 

Previously we used 1 big test contract and had to comment out functions in order to save time and be able to compile. Splitting up testing into separate folders allows for smaller testing contracts and no need to comment out stuff.

## Compile error for 2nd verification function (Fuel labs has been informed)
For verification, there are 2 algorithms; 1 that takes the public key and 1 that takes the secret scalar. However, Sway can't handle the latter function. This has been communicated to the Fuel labs team already. In the code this is indicated as well. We've tested the content of that function separately, so it can be uncommented as soon as Fuel labs has fixed the issue. 

## Running tests is slow
Running tests for verification and signing is very slow. For example, running the 2 tests in `verifying_test` takes 815.83s. This has been communicated to Fuel labs as well.

### CODE REVIEW

**Review by Mikerah — CHANGES_REQUESTED** [review]
- For functions that implement the low-level, non misuse resistant API from RustCrypto, there should be extra validation on the inputs to ensure it is not possible to forge or pass and otherwise invalid ecdsa signature. Further, it would be good to make this kind of internal and then have a higher level wrapper that abstracts it further.
- Since we don't want to expose any secret data on chain, we should remove `signingkey.sw`, `secretkey.sw` and `hmac.sw`

**Comment by Mikerah** [comment]
@ewynx The wrapper should probably emulate what is being done in the RustCrypto libraries. These wrappers should also have the extra validation needed as well.

**Comment by ewynx** [comment]
@Mikerah this is the ref impl we used https://github.com/RustCrypto/signatures/blob/master/ecdsa/src/hazmat.rs#L160 We don't see any more checks/wrappers to implement

**Comment by man2706kum** [comment]
The necessary validation checks has been added according to https://www.secg.org/sec1-v2.pdf and bitcoin ecdsa. The only wrapper for `verify_prehashed` is the `verify_prehash` in rustCcrypto which uses `bits2field`(https://github.com/RustCrypto/signatures/blob/ab5150ca46f2381c7cf5a4d5bfafcfb9c9085811/ecdsa/src/verifying.rs#LL163C17-L163C17). we don't need to use this because the scalar size and sha256 size both are 256 bit.

File: p256/src/ecdsa.sw [hunk 1]
```diff
@@ -12,13 +12,52 @@ pub struct Signature {
...
+  // same check for s.
+  assert_neq(sig.r.ls, Scalar::zero().ls);
+  assert_neq(sig.s.ls, Scalar::zero().ls);
+  assert_eq(mod_r.ls, sig.r.ls);
+  assert_eq(mod_s.ls, sig.s.ls);
+
+  //cheks if hash is non-zero
+  let mut i = 0;
+  while i < 32 {
+    if bytes[i] != 0 {
```
Comment (Mikerah) [inline p256/src/ecdsa.sw:50]: Is this constant time? I have a feeling that this loop isn't constant time

**Review by Mikerah — CHANGES_REQUESTED** [review]

**Review by Mikerah — APPROVED** [review]

**Pull request merged**